encodings/datetime-parts/src/compute/take.rs

```diff
@@ -1,22 +1,133 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: Copyright the Vortex contributors
 
-use vortex_array::compute::{TakeKernel, TakeKernelAdapter, take};
-use vortex_array::{Array, ArrayRef, IntoArray, register_kernel};
-use vortex_error::VortexResult;
+use vortex_array::compute::{TakeKernel, TakeKernelAdapter, fill_null, take};
+use vortex_array::{Array, ArrayRef, IntoArray, ToCanonical, register_kernel};
+use vortex_error::{VortexExpect, VortexResult, vortex_panic};
+use vortex_scalar::Scalar;
 
 use crate::{DateTimePartsArray, DateTimePartsVTable};
 
 impl TakeKernel for DateTimePartsVTable {
     fn take(&self, array: &DateTimePartsArray, indices: &dyn Array) -> VortexResult<ArrayRef> {
-        Ok(DateTimePartsArray::try_new(
-            array.dtype().clone(),
-            take(array.days(), indices)?,
-            take(array.seconds(), indices)?,
-            take(array.subseconds(), indices)?,
-        )?
-        .into_array())
+        // we go ahead and canonicalize here to avoid worst-case canonicalizing 3 separate times
+        let indices = indices.to_primitive()?;
+
+        let taken_days = take(array.days(), indices.as_ref())?;
+        let taken_seconds = take(array.seconds(), indices.as_ref())?;
+        let taken_subseconds = take(array.subseconds(), indices.as_ref())?;
+
+        // Update the dtype if the nullability changed due to nullable indices
+        let dtype = if taken_days.dtype().is_nullable() != array.dtype().is_nullable() {
+            array
+                .dtype()
+                .with_nullability(taken_days.dtype().nullability())
+        } else {
+            array.dtype().clone()
+        };
+
+        // DateTimePartsArray requires seconds and subseconds to be non-nullable.
+        // If they became nullable due to nullable indices, we need to fill nulls.
+        let (taken_seconds, taken_subseconds) = if taken_seconds.dtype().is_nullable() {
+            if !taken_days.dtype().is_nullable() {
+                vortex_panic!("Mismatched types: days is not nullable, seconds is nullable");
+            }
+            if !taken_subseconds.dtype().is_nullable() {
+                vortex_panic!("Mismatched types: seconds is nullable, subseconds is not");
+            }
+            if !indices.dtype().is_nullable() {
+                vortex_panic!("Mismatched types: indices are not nullable, seconds is nullable");
+            }
+
+            // get the first non-null index, and use that to fill both seconds and subseconds
```

Comment:
I think you should fill with 0 here to avoid worst case scenario of going through the whole array

Reply:
compromised: try getting Stat::Min if present (to help compression), fallback to 0